### ISSUE TITLE
Incorrect Shape Assertion and Unexpected Output Shapes in TabNetPretrainer Example

### ISSUE THREAD

**Issue by fbia** [dreamquark-ai/tabnet#571]
**Describe the bug**

There seems to be an issue with the [pretraining_example.ipynb](https://github.com/dreamquark-ai/tabnet/blob/develop/pretraining_example.ipynb) notebook. The assertion comparing the shapes of `reconstructed_X` and `embedded_X` appears incorrect, and attempting to assert that `reconstructed_X` has the same shape as `X_valid` also raises an error.

**What is the current behavior?**
The notebook contains the following assertion after the pretraining prediction:

```python
reconstructed_X, embedded_X = unsupervised_model.predict(X_valid)
assert(reconstructed_X.shape==embedded_X.shape)
```
This assertion incorrectly assumes that the reconstructed data and the embedded data should have the same shape.
However, when I replace it with:

```python
assert reconstructed_X.shape == X_valid.shape
```
I encounter an error, indicating that `reconstructed_X` does not match the shape of `X_valid`. This behavior is unexpected because `reconstructed_X` is supposed to be a reconstruction of the input data.

**If the current behavior is a bug, please provide the steps to reproduce.**

1. Open the [pretraining_example.ipynb](https://github.com/dreamquark-ai/tabnet/blob/develop/pretraining_example.ipynb) notebook.
2. Run all cells up to and including the cell with:
```python
reconstructed_X, embedded_X = unsupervised_model.predict(X_valid)
```
3. Replace the assertion with:
```python
assert reconstructed_X.shape == X_valid.shape
```
Observe that this raises an error.

**Expected behavior**

The expected behavior is:

`reconstructed_X` should have the same shape as `X_valid`, as it represents the model's attempt to reconstruct the input.

`embedded_X` should have a different shape (typically fewer features), as it represents a lower-dimensional embedding of the input.

If this is not how the method is intended to behave, clarification in the documentation or example notebook would be helpful.

The corrected assertions might look like this (if this behavior is confirmed to be correct):

```python
assert reconstructed_X.shape == X_valid.shape  # Reconstruction matches input
assert embedded_X.shape[0] == X_valid.shape[0]  # Same number of samples
assert embedded_X.shape[1] < X_valid.shape[1]  # Fewer features in embedding
```

**Screenshots**

**Other relevant information:**
poetry version:  
python version: 
Operating System: 
Additional tools:

**Additional context**

This issue could confuse new users trying to understand how `TabNetPretrainer` works. It would be great if you could clarify:
The expected relationship between `reconstructed_X`, `embedded_X`, and `X_valid`.

Whether there’s a bug in how shapes are handled or if this example needs correction.

Thank you for maintaining such a great library!

**Comment by Optimox** [comment]
Actually this notebook runs inside our CI, so everything should be running just fine...

**Comment by fbia** [comment]
I apologize if my initial explanation was unclear. The notebook itself functions correctly. My concern lies with the interpretation of the outputs of:
```python
reconstructed_X, embedded_X = unsupervised_model.predict(X_valid)
```

Currently, the code includes this check:
```python
assert(reconstructed_X.shape == embedded_X.shape)
```

My Concern:


The current assertion `assert(reconstructed_X.shape == embedded_X.shape)` suggests that `reconstructed_X` is the reconstruction of `embedded_X` (the embedded representation), rather than a reconstruction of the original input data `X_valid`.


Given that TabNet often involves self-supervised pretraining with a reconstruction task (where the model attempts to reconstruct the original input features), I would expect `reconstructed_X` to have the same shape as `X_valid`.

I propose this check instead:
```python
assert reconstructed_X.shape == X_valid.shape
```

This would verify that `reconstructed_X` represents the reconstruction of the original input data `X_valid`, as I understand the reconstruction task to be.

Question:

Is my understanding of the reconstruction task in this TabNet implementation correct? Or is there a specific reason why `reconstructed_X` is expected to have the same shape as `embedded_X`?

Thank you for your time and clarifications.

**Comment by Optimox** [comment]
Hi,
Yes pretraining optimize the reconstruction at the embedding level. This is because the loss would be much more complicated if you had to split continuous target and categorical ones. Also learning to reconstruct the exact inputs (last layer for reconstruction) would be a waste of compute since after pretraining you'll overwrite the reconstruction layer for your classication or regression layer.
The reconstruction task happens at the embedding level and is also made difficult by the ghost batch norm: you need to reconstruct inputs that are dependent of other inputs inside the batch. That is one limitation of the approach in my opinion, but that is what was used in the original paper.
